File: pdu/python/ev3_msgs/pdu_cdr_conv_Ev3PduSensor.py
from .pdu_pytype_Ev3PduSensor import Ev3PduSensor
from ..pdu_cdr_runtime import CdrReader, CdrWriter

from ..ev3_msgs.pdu_cdr_conv_Ev3PduColorSensor import *
from ..ev3_msgs.pdu_cdr_conv_Ev3PduSensorHeader import *
from ..ev3_msgs.pdu_cdr_conv_Ev3PduTouchSensor import *



def py_to_cdr_body_Ev3PduSensor(writer: CdrWriter, src: Ev3PduSensor):
    py_to_cdr_body_Ev3PduSensorHeader(writer, src.head)
    values = src.buttons
    for i in range(1):
        writer.write_uint8(values[i] if i < len(values) else 0)
    values = src.color_sensors
    for i in range(2):
        py_to_cdr_body_Ev3PduColorSensor(writer, values[i] if i < len(values) else Ev3PduColorSensor())
    values = src.touch_sensors
    for i in range(2):
        py_to_cdr_body_Ev3PduTouchSensor(writer, values[i] if i < len(values) else Ev3PduTouchSensor())
    values = src.motor_angle
    for i in range(3):
        writer.write_uint32(values[i] if i < len(values) else 0)
    writer.write_int32(src.gyro_degree)
    writer.write_int32(src.gyro_degree_rate)
    writer.write_uint32(src.sensor_ultrasonic)
    writer.write_float64(src.gps_lat)
    writer.write_float64(src.gps_lon)


def cdr_body_to_py_Ev3PduSensor(reader: CdrReader, dst: Ev3PduSensor):
    cdr_body_to_py_Ev3PduSensorHeader(reader, dst.head)
    dst.buttons = [reader.read_uint8() for _ in range(1)]
    dst.color_sensors = []
    for _ in range(2):
        elem = Ev3PduColorSensor()
        cdr_body_to_py_Ev3PduColorSensor(reader, elem)
        dst.color_sensors.append(elem)
    dst.touch_sensors = []
    for _ in range(2):
        elem = Ev3PduTouchSensor()
        cdr_body_to_py_Ev3PduTouchSensor(reader, elem)
        dst.touch_sensors.append(elem)
    dst.motor_angle = [reader.read_uint32() for _ in range(3)]
    dst.gyro_degree = reader.read_int32()
    dst.gyro_degree_rate = reader.read_int32()
    dst.sensor_ultrasonic = reader.read_uint32()
    dst.gps_lat = reader.read_float64()
    dst.gps_lon = reader.read_float64()
    return dst


def py_to_cdr_Ev3PduSensor(src: Ev3PduSensor) -> bytes:
    writer = CdrWriter()
    writer.write_encapsulation()
    py_to_cdr_body_Ev3PduSensor(writer, src)
    return writer.bytes()


def cdr_to_py_Ev3PduSensor(cdr_payload) -> Ev3PduSensor:
    reader = CdrReader(cdr_payload)
    reader.read_encapsulation()
    dst = Ev3PduSensor()
    return cdr_body_to_py_Ev3PduSensor(reader, dst)

File: pdu/javascript/ev3_msgs/pdu_cdr_conv_Ev3PduSensor.js
import { Ev3PduSensor } from './pdu_jstype_Ev3PduSensor.js';
import { PduCdrWriter, PduCdrReader } from '../pdu_cdr_runtime.js';
import { Ev3PduColorSensor } from '../ev3_msgs/pdu_jstype_Ev3PduColorSensor.js';
import { PduEv3PduColorSensorConverter } from '../ev3_msgs/pdu_cdr_conv_Ev3PduColorSensor.js';
import { Ev3PduSensorHeader } from '../ev3_msgs/pdu_jstype_Ev3PduSensorHeader.js';
import { PduEv3PduSensorHeaderConverter } from '../ev3_msgs/pdu_cdr_conv_Ev3PduSensorHeader.js';
import { Ev3PduTouchSensor } from '../ev3_msgs/pdu_jstype_Ev3PduTouchSensor.js';
import { PduEv3PduTouchSensorConverter } from '../ev3_msgs/pdu_cdr_conv_Ev3PduTouchSensor.js';


export class PduEv3PduSensorConverter {
    /**
     * @param {PduCdrWriter} writer
     * @param { Ev3PduSensor } src
     */
    static to_cdr_body(writer, src) {
        PduEv3PduSensorHeaderConverter.to_cdr_body(writer, src.head);
        {
            const values = src.buttons;
            for (let i = 0; i < 1; i++) {
                writer.write_uint8(i < values.length ? values[i] : 0);
            }
        }
        {
            const values = src.color_sensors;
            for (let i = 0; i < 2; i++) {
                PduEv3PduColorSensorConverter.to_cdr_body(writer, i < values.length ? values[i] : new Ev3PduColorSensor());
            }
        }
        {
            const values = src.touch_sensors;
            for (let i = 0; i < 2; i++) {
                PduEv3PduTouchSensorConverter.to_cdr_body(writer, i < values.length ? values[i] : new Ev3PduTouchSensor());
            }
        }
        {
            const values = src.motor_angle;
            for (let i = 0; i < 3; i++) {
                writer.write_uint32(i < values.length ? values[i] : 0);
            }
        }
        writer.write_int32(src.gyro_degree);
        writer.write_int32(src.gyro_degree_rate);
        writer.write_uint32(src.sensor_ultrasonic);
        writer.write_float64(src.gps_lat);
        writer.write_float64(src.gps_lon);
    }

    /**
     * @param {PduCdrReader} reader
     * @param { Ev3PduSensor } dst
     * @returns { Ev3PduSensor }
     */
    static cdr_body_to_js(reader, dst) {
        PduEv3PduSensorHeaderConverter.cdr_body_to_js(reader, dst.head);
        dst.buttons = Array.from({ length: 1 }, () => reader.read_uint8());
        dst.color_sensors = [];
        for (let i = 0; i < 2; i++) {
            dst.color_sensors.push(PduEv3PduColorSensorConverter.cdr_body_to_js(reader, new Ev3PduColorSensor()));
        }
        dst.touch_sensors = [];
        for (let i = 0; i < 2; i++) {
            dst.touch_sensors.push(PduEv3PduTouchSensorConverter.cdr_body_to_js(reader, new Ev3PduTouchSensor()));
        }
        dst.motor_angle = Array.from({ length: 3 }, () => reader.read_uint32());
        dst.gyro_degree = reader.read_int32();
        dst.gyro_degree_rate = reader.read_int32();
        dst.sensor_ultrasonic = reader.read_uint32();
        dst.gps_lat = reader.read_float64();
        dst.gps_lon = reader.read_float64();
        return dst;
    }

    /**
     * @param { Ev3PduSensor } src
     * @returns {ArrayBuffer}
     */
    static to_cdr(src) {
        const writer = new PduCdrWriter();
        writer.write_encapsulation();
        this.to_cdr_body(writer, src);
        return writer.get_buf();
    }

    /**
     * @param {ArrayBuffer|ArrayBufferView} cdrPayload
     * @returns { Ev3PduSensor }
     */
    static from_cdr(cdrPayload) {
        const reader = new PduCdrReader(cdrPayload);
        reader.read_encapsulation();
        return this.cdr_body_to_js(reader, new Ev3PduSensor());
    }
}
